### PR TITLE
Feature/tooltip closing

### PR DESCRIPTION
**Pull request by JoaquinPalacios**
# Implement tooltip close on outside click for trigger="click"

## Description

This PR implements the ability to close click-triggered tooltips by clicking outside the chart area. Previously, tooltips with trigger="click" would remain open until the user clicked on another chart element or manually closed them. Now, clicking anywhere outside the chart will close the tooltip.

## Related Issue

#3573 

## Motivation and Context

Users expect click-triggered tooltips to behave like standard UI tooltips that close when clicking outside their area. This is a common UX pattern that improves usability and provides a more intuitive interaction model for charts with click-triggered tooltips.

## How Has This Been Tested?

- Unit Tests: Added comprehensive tests in tooltipEventType.spec.tsx to verify outside click behavior for both axis-based and item-based tooltips
- Integration Tests: Verified that the feature works across all chart types (Bar, Pie, Scatter, Funnel, RadialBar, Treemap)
- Regression Testing: Ensured existing tooltip functionality remains intact for hover-triggered tooltips
- Edge Cases: Tested scenarios with multiple charts, external event handlers, and various tooltip configurations

### Test Environment:

- Node.js environment with Jest test runner
- All existing tooltip tests continue to pass
- New outside click tests specifically validate the new functionality

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes

## Note: The implementation includes:

- Redux actions for tooltip closing (tooltipCloseStart, tooltipCloseEnd, clearClickTooltip)
- Outside click detection in RechartsWrapper component
- Updated event handler logic in chart components to properly forward external handlers
- Fixed Treemap rendering issue that was discovered during testing
- Comprehensive test coverage for the new functionality
- The changes are backward compatible and only affect click-triggered tooltips, leaving hover-triggered tooltips unchanged.

### CODE REVIEW

File: src/cartesian/Bar.tsx [hunk 1]
```diff
@@ -242,28 +256,73 @@ function BarBackground(props: BarBackgroundProps) {
...
         }
 
-        // @ts-expect-error BarRectangleItem type definition says it's missing properties, but I can see them present in debugger!
-        const onMouseEnter = onMouseEnterFromContext(entry, i);
-        // @ts-expect-error BarRectangleItem type definition says it's missing properties, but I can see them present in debugger!
-        const onMouseLeave = onMouseLeaveFromContext(entry, i);
-        // @ts-expect-error BarRectangleItem type definition says it's missing properties, but I can see them present in debugger!
-        const onClick = onClickFromContext(entry, i);
+        // Always attach external event handlers (passed as props)
+        const eventHandlers: any = {
```
Comment (ckifer) [inline src/cartesian/Bar.tsx]: can this be typed?

File: src/cartesian/Bar.tsx [hunk 2]
```diff
@@ -242,28 +256,73 @@ function BarBackground(props: BarBackgroundProps) {
...
+          // For hover mode, also ensure external onClick is called if provided
+          const externalOnClick = eventHandlers.onClick;
+          if (externalOnClick) {
+            eventHandlers.onClick = (e: React.MouseEvent<SVGElement, MouseEvent>) => {
+              externalOnClick(entry, i, e);
+            };
+          }
+        }
+
+        const noop = () => {}; // function to prevent errors when event handlers are undefined
```
Comment (ckifer) [inline src/cartesian/Bar.tsx]: I think we have a noop function defined that can be imported

File: src/cartesian/Bar.tsx [hunk 3]
```diff
@@ -329,16 +397,62 @@ function BarRectangles({
           dataKey,
         };
 
+        // Always attach external event handlers (passed as props)
+        const eventHandlers: any = {
+          ...adaptEventsOfChild(restOfAllOtherProps, entry, i),
+        };
+
+        // Conditionally attach internal tooltip handlers based on tooltip trigger
+        if (tooltipTrigger === 'click') {
```
Comment (ckifer) [inline src/cartesian/Bar.tsx]: I'm confused at what all of this is doing that what was here wasn't doing?

these three events are already combined internal and external using `useMouseClickItemDispatch`, etc.

File: src/chart/RechartsWrapper.tsx [hunk 4]
```diff
@@ -194,7 +226,7 @@ export const RechartsWrapper = forwardRef(
             onTouchStart={myOnTouchStart}
             ref={innerRef}
           >
-            {children}
+            {React.isValidElement(children) ? React.cloneElement(children) : children}
```
Comment (ckifer) [inline src/chart/RechartsWrapper.tsx]: why is this needed? We're trying to avoid adding more cloning

File: src/chart/Treemap.tsx [hunk 5]
```diff
@@ -382,6 +382,10 @@ type ContentItemProps = {
   onClick?: (e: React.MouseEvent<SVGPathElement, MouseEvent>) => void;
   onMouseEnter?: (e: React.MouseEvent<SVGPathElement, MouseEvent>) => void;
   onMouseLeave?: (e: React.MouseEvent<SVGPathElement, MouseEvent>) => void;
+  externalOnClick?: (node: TreemapNode, e: React.MouseEvent<SVGPathElement, MouseEvent>) => void;
+  externalOnMouseEnter?: (node: TreemapNode, e: React.MouseEvent<SVGPathElement, MouseEvent>) => void;
+  externalOnMouseLeave?: (node: TreemapNode, e: React.MouseEvent<SVGPathElement, MouseEvent>) => void;
+  internalOnClick?: (node: TreemapNode) => void;
```
Comment (ckifer) [inline src/chart/Treemap.tsx:388]: why do we need these? There are already events here?

File: src/polar/Pie.tsx [hunk 6]
```diff
@@ -456,17 +465,63 @@ function PieSectors(props: PieSectorsProps) {
           [DATA_ITEM_DATAKEY_ATTRIBUTE_NAME]: allOtherPieProps.dataKey,
         };
 
+        // Always attach external event handlers (passed as props)
+        const eventHandlers: any = {
```
Comment (ckifer) [inline src/polar/Pie.tsx]: this same logic is duplicated in 6-ish places, it can probably moved to a single function

File: src/polar/Pie.tsx [hunk 7]
```diff
@@ -456,17 +465,63 @@ function PieSectors(props: PieSectorsProps) {
...
+          ...adaptEventsOfChild(restOfAllOtherProps, entry, i),
+        };
+
+        // Conditionally attach internal tooltip handlers based on tooltip trigger
+        if (tooltipTrigger === 'click') {
+          const triggerInfo = {
+            tooltipPayload: entry.tooltipPayload as any,
```
Comment (ckifer) [inline src/polar/Pie.tsx]: why do we need to cast and why to any?

File: src/state/tooltipSlice.ts [hunk 8]
```diff
@@ -219,6 +219,11 @@ export type TooltipState = {
    * we have to change this to an array - and update all the places reading this state too.
    */
   settings: TooltipSettingsState;
+  /**
+   * This is a flag that is set when the tooltip is closing.
+   * This is used to prevent the tooltip from being closed again.
+   */
```
Comment (ckifer) [inline src/state/tooltipSlice.ts]: Can't we just check state and derive that the tooltip is not showing?

File: src/chart/RechartsWrapper.tsx [hunk 9]
```diff
@@ -51,11 +51,42 @@ export const RechartsWrapper = forwardRef(
...
     useSynchronisedEventsFromOtherCharts();
 
     const setScaleRef = useReportScale();
 
+    const handleRequestTooltipClose = useCallback(() => {
+      dispatch(tooltipCloseStart());
+      setTimeout(() => {
+        dispatch(tooltipCloseEnd());
+      }, 100);
+    }, [dispatch]);
```
Comment (ckifer) [inline src/chart/RechartsWrapper.tsx]: I don't think we want to introduce timeouts if we can help it, this seems like something that might cause some funky bugs in the long run

File: src/chart/RechartsWrapper.tsx [hunk 10]
```diff
@@ -51,11 +51,42 @@ export const RechartsWrapper = forwardRef(
...
+        if (chartRef.current && !chartRef.current.contains(event.target as Node)) {
+          handleRequestTooltipClose();
+        }
+      };
+
+      document.addEventListener('mousedown', handleDocumentClick, true);
+      return () => {
+        document.removeEventListener('mousedown', handleDocumentClick, true);
+      };
+    }, [hasClickTooltipActive, handleRequestTooltipClose]);
```
Comment (ckifer) [inline src/chart/RechartsWrapper.tsx]: I think if this is the only thing we did in this PR the concern would be solved right?

we can request the tooltip to be closed if its open by calling `dispatch` and updating tooltip state directly? No need to update all of the event handlers

File: src/state/tooltipSlice.ts [hunk 11]
```diff
@@ -325,6 +333,32 @@ const tooltipSlice = createSlice({
...
+       */
+      state.isClosing = true;
+      state.itemInteraction.click.active = false;
+      state.axisInteraction.click.active = false;
+    },
+    tooltipCloseEnd: state => {
+      /*
+       * End the tooltip closing process.
+       * This action resets the isClosing flag to false, allowing new tooltip interactions.
+       * Should be called after a short delay to prevent race conditions.
```
Comment (ckifer) [inline src/state/tooltipSlice.ts]: Its adding a race condition I think

File: test/chart/Treemap.spec.tsx [hunk 12]
```diff
@@ -4,11 +4,188 @@ import { render, fireEvent } from '@testing-library/react';
...
+    const { container } = render(
+      <svg>
+        <Rectangle x={0} y={0} width={100} height={100} />
+      </svg>,
+    );
+
+    expect(container.querySelector('.recharts-rectangle')).toBeInTheDocument();
+  });
+
+  test('should render Animate component directly', () => {
```
Comment (ckifer) [inline test/chart/Treemap.spec.tsx]: there's a bunch of random tests in here that don't apply to Treemap at all?

File: src/cartesian/Bar.tsx (hunk 3 above)
Comment (JoaquinPalacios) [reply src/cartesian/Bar.tsx]: Thanks for the feedback @ckifer you are right, the code was redundant. I have just committed a refactor of it.

File: src/chart/RechartsWrapper.tsx (hunk 4 above)
Comment (JoaquinPalacios) [reply src/chart/RechartsWrapper.tsx]: I have removed it. At the beginning of this work, I initially thought the React.cloneElement(children) was necessary for some edge cases, but I didn't go back to verify if it was still needed after completing the implementation. My bad - I should have double-checked this before submitting the PR.

File: src/cartesian/Bar.tsx (hunk 2 above)
Comment (JoaquinPalacios) [reply src/cartesian/Bar.tsx]: Thanks, I have now import it from `es-toolkit`

File: src/chart/Treemap.tsx (hunk 5 above)
Comment (JoaquinPalacios) [reply src/chart/Treemap.tsx:388]: Hi @ckifer  these event handler props are necessary to make sure that the user-provided event handlers are actually invoked when the relevant DOM events occur on the rendered nodes, while still allowing the component to run its own internal logic for those events. I had to come up with this solution otherwise the test "navigates through nested nodes correctly" in line 220 of Treemap.spec.tsx fails because no click handler exist (internalOnClick) and it had no navigation logic (the Treemap doesn't change its view when clicked).

What would you advise me to do in this case? Keep the external event handlers to maintain the component's public API and navigation functionality or remove the navigation test? Perhaps I am missing something else. Let me know, thanks.

File: src/polar/Pie.tsx (hunk 6 above)
Comment (JoaquinPalacios) [reply src/polar/Pie.tsx]: Hi @ckifer thanks for your feedback, I've refactored this duplicated event handler logic into a reusable createEventHandlers utility function in ReactUtils.ts. I have just done a commit with the updates.

File: src/polar/Pie.tsx (hunk 7 above)
Comment (JoaquinPalacios) [reply src/polar/Pie.tsx]: This has been resolved on the commit 35036e2

File: src/cartesian/Bar.tsx (hunk 1 above)
Comment (JoaquinPalacios) [reply src/cartesian/Bar.tsx]: This has been resolved on the commit https://github.com/recharts/recharts/commit/35036e2756626c55f08a8c24b0eba18e00637b3f

File: src/state/tooltipSlice.ts (hunk 8 above)
Comment (JoaquinPalacios) [reply src/state/tooltipSlice.ts]: The flag prevents race conditions during the brief window when a tooltip is being closed but the middleware hasn't finished processing the close event. I have updated it the comment on a new commit https://github.com/recharts/recharts/pull/6007/commits/c6d62443b219e8872a663331f09dde1919daea0b . Please let me know if you consider that I should seek a different solution.

File: src/chart/RechartsWrapper.tsx (hunk 9 above)
Comment (JoaquinPalacios) [reply src/chart/RechartsWrapper.tsx]: Thanks for the feedback, I completely agree about the bugs it could introduce in the long term.

I've updated the PR with a new approach that removes the timer entirely. Now, the mouseMove event immediately following a close action is caught by the middleware, which then "consumes" a gate flag and resets the state. This is now on this new commit [5b5258a](https://github.com/recharts/recharts/pull/6007/commits/5b5258acfd6cf374f9aac95053db9277e2e79504)

File: src/chart/RechartsWrapper.tsx (hunk 10 above)
Comment (JoaquinPalacios) [reply src/chart/RechartsWrapper.tsx]: Thanks for the suggestion! I have just made a commit [3787ce4](https://github.com/recharts/recharts/pull/6007/commits/3787ce4b94ab5e9b855a15397406b095fec2389d) with a more simplify approach.

**Comment by ckifer** [comment]
Will take a look at the changes tomorrow my time, thanks for making them!

File: src/state/tooltipSlice.ts (hunk 11 above)
Comment (JoaquinPalacios) [reply src/state/tooltipSlice.ts]: Now we the addition of the previous work, I've pushed a new commit that removes the isClosing state and the tooltipCloseStart/End actions entirely. The fix now exclusively uses the existing clearClickTooltip action, which is dispatched directly by the middleware for "tap away" and the useEffect for "outside clicks."

This makes the solution much simpler and more direct.

File: src/chart/RechartsWrapper.tsx [hunk 13]
```diff
@@ -51,11 +51,42 @@ export const RechartsWrapper = forwardRef(
...
 
     useSynchronisedEventsFromOtherCharts();
 
     const setScaleRef = useReportScale();
 
+    const handleRequestTooltipClose = useCallback(() => {
+      dispatch(tooltipCloseStart());
+      setTimeout(() => {
+        dispatch(tooltipCloseEnd());
+      }, 100);
```
[the remark this answers is not in the document]
Comment (JoaquinPalacios) [reply src/chart/RechartsWrapper.tsx]: This has been addressed on a commit.

File: test/chart/Treemap.spec.tsx (hunk 12 above)
Comment (JoaquinPalacios) [reply test/chart/Treemap.spec.tsx]: Sorry, my bad. On one of first commits it failed to cover the percentage of code added, I added some tests and because eas still failing I let cursor to add more and clearly didn't help much. I have removed them entirely from this file.

File: src/chart/RechartsWrapper.tsx [hunk 14]
```diff
@@ -51,11 +51,42 @@ export const RechartsWrapper = forwardRef(
...
+
+      const handleDocumentClick = (event: MouseEvent) => {
+        if (chartRef.current && !chartRef.current.contains(event.target as Node)) {
+          handleRequestTooltipClose();
+        }
+      };
+
+      document.addEventListener('mousedown', handleDocumentClick, true);
+      return () => {
+        document.removeEventListener('mousedown', handleDocumentClick, true);
```
[the remark this answers is not in the document]
Comment (JoaquinPalacios) [reply src/chart/RechartsWrapper.tsx]: Added 'pointerdown' on the latest commit

**Comment by JoaquinPalacios** [comment]
Hi @ckifer thanks for all your feedback and thorough check of the PR. I ran into some original errors while trying to implement this feature that led to the addition of unnecessary code and a bit of overuse of Cursor. I’ve now addressed most of the comments and cleaned up the repeated code and tests. There is one question for you on one of the comments. Please let me know about it and if there is anything else I should look into it.

File: src/cartesian/Bar.tsx [hunk 15]
```diff
@@ -243,28 +251,34 @@ function BarBackground(props: BarBackgroundProps) {
...
-          onMouseEnter,
-          onMouseLeave,
-          onClick,
+          ...eventHandlers,
           dataKey,
           index: i,
           className: 'recharts-bar-background-rectangle',
+          onClick: eventHandlers.onClick || noop,
+          onMouseEnter: eventHandlers.onMouseEnter || noop,
+          onMouseLeave: eventHandlers.onMouseLeave || noop,
```
Comment (ckifer) [inline src/cartesian/Bar.tsx]: nit: can you default these to `noop` in `createEventHandlers` so that we don't have to import it in every file?

File: src/cartesian/Bar.tsx [hunk 16]
```diff
@@ -243,28 +251,34 @@ function BarBackground(props: BarBackgroundProps) {
           return null;
         }
 
-        // @ts-expect-error BarRectangleItem type definition says it's missing properties, but I can see them present in debugger!
-        const onMouseEnter = onMouseEnterFromContext(entry, i);
-        // @ts-expect-error BarRectangleItem type definition says it's missing properties, but I can see them present in debugger!
-        const onMouseLeave = onMouseLeaveFromContext(entry, i);
-        // @ts-expect-error BarRectangleItem type definition says it's missing properties, but I can see them present in debugger!
-        const onClick = onClickFromContext(entry, i);
+        // Use the new utility function
```
Comment (ckifer) [inline src/cartesian/Bar.tsx]: nit nit: this won't always be new, can probably remove these comments

File: src/chart/Treemap.tsx [hunk 17]
```diff
@@ -793,7 +820,7 @@ class TreemapWithState extends PureComponent<InternalTreemapProps, State> {
     const { formatRoot } = this.state;
 
     if (!formatRoot) {
-      return null;
+      return <g />;
```
Comment (ckifer) [inline src/chart/Treemap.tsx]: why change what is returned here? Can we change this back to `null`?

File: src/chart/Treemap.tsx [hunk 18]
```diff
@@ -676,22 +701,17 @@ class TreemapWithState extends PureComponent<InternalTreemapProps, State> {
...
     const { width, height, x, y, depth } = nodeProps;
     const translateX = parseInt(`${(Math.random() * 2 - 1) * width}`, 10);
-    let event = {} as any;
-    if (isLeaf || type === 'nest') {
-      event = {
-        onMouseEnter: this.handleMouseEnter.bind(this, nodeProps),
-        onMouseLeave: this.handleMouseLeave.bind(this, nodeProps),
-        onClick: this.handleClick.bind(this, nodeProps),
-      };
-    }
```
Comment (ckifer) [inline src/chart/Treemap.tsx:690]: doesn't this break nesting functionality?

File: src/chart/Treemap.tsx [hunk 19]
```diff
@@ -433,14 +437,11 @@ function ContentItem({
 
   const colors = colorPanel || COLOR_PANEL;
   return (
-    <g>
+    <g onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} onClick={onClick}>
```
Comment (ckifer) [inline src/chart/Treemap.tsx:440]: why move these to the group?

File: src/chart/Treemap.tsx (hunk 5 above)
Comment (ckifer) [reply src/chart/Treemap.tsx:388]: trying to understand if these are being exposed in the public interface or not. If so, we can't do this

File: src/polar/Pie.tsx [hunk 20]
```diff
@@ -743,46 +758,54 @@ const defaultPieProps = {
 function PieImpl(props: Props) {
   const propsWithDefaults: InternalProps = resolveDefaultProps(props, defaultPieProps);
 
+  // Merge event handlers into propsWithDefaults
+  const mergedProps: InternalProps = {
+    ...propsWithDefaults,
+    onClick: props.onClick,
+    onMouseEnter: props.onMouseEnter,
+    onMouseLeave: props.onMouseLeave,
```
Comment (ckifer) [inline src/polar/Pie.tsx]: do these need included here? I think they just get filtered out below?

File: src/util/ReactUtils.ts [hunk 21]
```diff
@@ -176,3 +187,34 @@ export const filterProps = (
...
+ */
+export const createEventHandlers = (
+  restOfAllOtherProps: any,
+  entry: Entry,
+  itemIndex: number,
+  triggerInfo: TriggerInfo,
+  contextHandlers: ContextHandlers,
+): any => {
+  // Always attach external event handlers (passed as props)
+  const eventHandlers: any = {
```
Comment (ckifer) [inline src/util/ReactUtils.ts:212]: Can this be typed?

File: test/component/Tooltip/Tooltip.visibility.spec.tsx [hunk 22]
```diff
@@ -227,7 +227,7 @@ const FunnelChartTestCase: TooltipVisibilityTestCase = {
     </FunnelChart>
   ),
   mouseHoverSelector: funnelChartMouseHoverTooltipSelector,
-  expectedTransform: 'transform: translate(355px, 55px);',
+  expectedTransform: 'transform: translate(40px, 15px);',
```
Comment (ckifer) [inline test/component/Tooltip/Tooltip.visibility.spec.tsx]: why did this change?

**Review by ckifer — COMMENTED** [review]
Looking better, thank you. I still want to pull this down and test it a bit. Added a few nits and questions

File: src/cartesian/Bar.tsx (hunk 1 above)
Comment (ckifer) [reply src/cartesian/Bar.tsx]: looks like this is still there, just in a different place

File: src/chart/Treemap.tsx (hunk 18 above)
Comment (PavelVanecek) [reply src/chart/Treemap.tsx:690]: Also why is this change here? How is it related to tooltip closing?

File: src/state/mouseEventsMiddleware.ts [hunk 23]
```diff
@@ -25,6 +28,24 @@ mouseClickMiddleware.startListening({
           activeCoordinate: activeProps.activeCoordinate,
         }),
       );
+    } else {
```
Comment (PavelVanecek) [inline src/state/mouseEventsMiddleware.ts:31]: This looks like the only relevant change to the PR title.

File: src/util/ReactUtils.ts (hunk 21 above)
Comment (PavelVanecek) [reply src/util/ReactUtils.ts:212]: Also can this be in separate PR?

File: test/component/Tooltip/Tooltip.visibility.spec.tsx (hunk 22 above)
Comment (PavelVanecek) [reply test/component/Tooltip/Tooltip.visibility.spec.tsx]: This looks like a bug to me. Why did this coordinate change? I do not see how it's related to the PR target.

File: test/state/mouseEventsMiddleware.spec.ts [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+      const state = store.getState().tooltip;
+      expect(state.itemInteraction.click.active).toBe(false);
+      expect(state.axisInteraction.click.active).toBe(false);
+    });
+
+    it('should activate axis tooltip when clicking on valid axis area', async () => {
+      store = createMinimalStore();
+
+      // Mock the selector to return valid active props
+      const mockSelectActivePropsFromChartPointer = vi.mocked(
```
Comment (PavelVanecek) [inline test/state/mouseEventsMiddleware.spec.ts]: Why so many mocks? Can we replace the mocks with actual interactions? That way we will not depend on the selector/state implementation.

**Review by PavelVanecek — CHANGES_REQUESTED** [review]
Little bit larger than I would like but okay. Can you please figure out the failing test, and mocks, before merging?

File: src/cartesian/Bar.tsx (hunk 15 above)
Comment (JoaquinPalacios) [reply src/cartesian/Bar.tsx]: I have just done this. See commit d80d13d

File: src/cartesian/Bar.tsx (hunk 16 above)
Comment (JoaquinPalacios) [reply src/cartesian/Bar.tsx]: removed it.

File: src/chart/Treemap.tsx (hunk 17 above)
Comment (JoaquinPalacios) [reply src/chart/Treemap.tsx]: Previously I had errors on the Treemap. It seems some of the changes I did a the beginning made me modify unnecessary code. I had put null back and works as expect it e7c3906

File: src/chart/Treemap.tsx (hunk 18 above)
Comment (JoaquinPalacios) [reply src/chart/Treemap.tsx:690]: No on my end, is there anything in particular I should test to dig further?

File: src/chart/Treemap.tsx (hunk 19 above)
Comment (JoaquinPalacios) [reply src/chart/Treemap.tsx:440]: Screenshot 2025-07-04 at 4 07 19 pm

If not, this test fails.

Comment (JoaquinPalacios) [reply src/chart/Treemap.tsx:440]: Please let me know if I should seek a different solution.

**Comment by JoaquinPalacios** [comment]
Hi @ckifer this is in response to [this](https://github.com/recharts/recharts/pull/6007#discussion_r2178719541) comment that I cannot reply below it. 

As per my understanding of the code the externalOnClick, externalOnMouseEnter, externalOnMouseLeave, and internalOnClick props are not part of the public API. They are internal props within the ContentItemWithEvents component.

File: src/polar/Pie.tsx (hunk 20 above)
Comment (JoaquinPalacios) [reply src/polar/Pie.tsx]: You're right, those lines are redundant. The event handlers are passed down through the rest of the props and picked up by createEventHandlers. I'd removed them on the commit [bbcc682](https://github.com/recharts/recharts/pull/6007/commits/bbcc682e20e1b08cf696543e42548dc71b2217ee)

File: src/util/ReactUtils.ts (hunk 21 above)
Comment (JoaquinPalacios) [reply src/util/ReactUtils.ts:212]: Hi @ckifer and @PavelVanecek I really tried to typed properly but I couldn't as the Bar.tsx expects one type and the rest another, and both won't accept to have both types on it. Happy to create another PR, question about how should I integrate this with my current PR as well?

File: test/component/Tooltip/Tooltip.visibility.spec.tsx (hunk 22 above)
Comment (JoaquinPalacios) [reply test/component/Tooltip/Tooltip.visibility.spec.tsx]: I had reverted back he changes and updated the trigerInfo function in the Funnel.tsx and solved the issue. The tooltip was being positioned at the top-left corner of the funnel segment instead of its center. I've corrected the logic to use the proper tooltipPosition, and the test now passes with the original expected value. See this commit 
1969150 about it.

File: test/state/mouseEventsMiddleware.spec.ts (hunk 24 above)
Comment (JoaquinPalacios) [reply test/state/mouseEventsMiddleware.spec.ts]: I have simplified and updated the test file with actual interactions on this commit [fd439be](https://github.com/recharts/recharts/pull/6007/commits/fd439bea49b33934c879d535ce727b52c3a2b0c4)

**Comment by ckifer** [comment]
I'm going to close this as stale, I think we can do this in a more elegant way without so many changes to events, etc.
